src/Features/Core/Portable/AddConstructorParametersFromMembers/AddConstructorParametersFromMembersCodeRefactoringProvider.State.cs

```diff
@@ -1,109 +1,140 @@
-                ImmutableArray<IParameterSymbol> parameters)
+                ImmutableArray<IParameterSymbol> parametersForSelectedMembers,
+                ImmutableArray<ISymbol> selectedMembers,
+                Document document)
             {
-                var parameterNames = parameters.SelectAsArray(p => p.Name);
-                return containingType.InstanceConstructors
-                    .Where(constructor => IsApplicableConstructor(constructor, parameterNames))
-                    .OrderByDescending(constructor => constructor.Parameters.Length)
-                    .FirstOrDefault();
+                var parameterNamesForSelectedMembers = parametersForSelectedMembers.SelectAsArray(p => p.Name);
+                var applicableConstructors = ArrayBuilder<ConstructorCandidate>.GetInstance();
+                var constructors = containingType.InstanceConstructors;
+                foreach (var constructor in constructors)
+                {
+                    var constructorParams = constructor.Parameters;
+
+                    if (constructorParams.Length == 2)
+                    {
+                        var compilation = await document.Project.GetCompilationAsync().ConfigureAwait(false);
+                        var deserializationConstructorCheck = new DeserializationConstructorCheck(compilation);
+                        if (deserializationConstructorCheck.IsDeserializationConstructor(constructor))
+                        {
+                            continue;
+                        }
+                    }
+
+                    if (!constructorParams.All(parameter => parameter.RefKind == RefKind.None) ||
+                        (constructorParams.Length == 0 && constructor.IsImplicitlyDeclared) ||
+                        constructorParams.Any(p => p.IsParams) ||
+                        SelectedMembersAlreadyExistAsParameters(parameterNamesForSelectedMembers, constructorParams))
+                    {
+                        continue;
+                    }
+
+                    var missingParametersBuilder = ArrayBuilder<IParameterSymbol>.GetInstance();
+                    var missingMembersBuilder = ArrayBuilder<ISymbol>.GetInstance();
+                    var constructorParamNames = constructor.Parameters.SelectAsArray(p => p.Name);
+                    var zippedParametersAndSelectedMembers = parametersForSelectedMembers.Zip(selectedMembers, (parameter, selectedMember) => (parameter, selectedMember));
+                    foreach ((var parameter, var selectedMember) in zippedParametersAndSelectedMembers)
+                    {
+                        if (!constructorParamNames.Contains(parameter.Name))
+                        {
+                            missingParametersBuilder.Add(parameter);
+                            missingMembersBuilder.Add(selectedMember);
+                        }
+                    }
+
+                    if (missingParametersBuilder != null)
```

Comment:
Since we are initializing `missingParametersBuilder` to an ArrayBuilder above, would this always be true? Should this be a Length check instead?

Reply:
Good point.  Also, I realized a check is not needed here.  Since `!SelectedMemersAlreadyExistAsParameters` was called above we know there will be missing parameters.